crates/ty_python_semantic/src/types/infer/builder/type_expression.rs

```diff
@@ -20,6 +20,18 @@ use crate::types::{
 impl<'db> TypeInferenceBuilder<'db, '_> {
     /// Infer the type of a type expression.
     pub(super) fn infer_type_expression(&mut self, expression: &ast::Expr) -> Type<'db> {
+        // `DeferredExpressionState::InStringAnnotation` takes precedence over other states.
+        // However, if it's not a stringified annotation, we must still ensure that annotation expressions
+        // are always deferred in stub files.
+        match self.deferred_state {
+            DeferredExpressionState::None => {
+                if self.in_stub() {
+                    self.deferred_state = DeferredExpressionState::Deferred;
+                }
+            }
+            DeferredExpressionState::InStringAnnotation(_) | DeferredExpressionState::Deferred => {}
+        }
```

Comment:
In general, setting the deferred state without a paired revert back to the previous deferred state is not right; it will cause the new deferred state to leak back up the stack, outside of this type expression inference.

I feel like there might be some refactor called for here between this method and `infer_type_expression_with_state`, but I haven't quite sorted yet what it should be...

Reply:
> I feel like there might be some refactor called for here between this method and `infer_type_expression_with_state`, but I haven't quite sorted yet what it should be...

Yeah, I also looked briefly at whether a refactor would be possible, but it seemed nontrivial...